hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieFileIndex.scala

```diff
@@ -122,81 +125,120 @@ case class HoodieFileIndex(spark: SparkSession,
     //    - Col-Stats Index is present
     //    - List of predicates (filters) is present
     val candidateFilesNamesOpt: Option[Set[String]] =
-      lookupCandidateFilesInMetadataTable(dataFilters) match {
-        case Success(opt) => opt
-        case Failure(e) =>
-          logError("Failed to lookup candidate files in File Index", e)
-
-          spark.sqlContext.getConf(DataSkippingFailureMode.configName, DataSkippingFailureMode.Fallback.value) match {
-            case DataSkippingFailureMode.Fallback.value => Option.empty
-            case DataSkippingFailureMode.Strict.value   => throw new HoodieException(e);
-          }
-      }
+    lookupCandidateFilesInMetadataTable(dataFilters) match {
+      case Success(opt) => opt
+      case Failure(e) =>
+        logError("Failed to lookup candidate files in File Index", e)
+
+        spark.sqlContext.getConf(DataSkippingFailureMode.configName, DataSkippingFailureMode.Fallback.value) match {
+          case DataSkippingFailureMode.Fallback.value => Option.empty
+          case DataSkippingFailureMode.Strict.value => throw new HoodieException(e);
+        }
+    }
 
     logDebug(s"Overlapping candidate files from Column Stats Index: ${candidateFilesNamesOpt.getOrElse(Set.empty)}")
 
-    var totalFileSize = 0
-    var candidateFileSize = 0
-
-    // Prune the partition path by the partition filters
-    // NOTE: Non-partitioned tables are assumed to consist from a single partition
-    //       encompassing the whole table
-    val prunedPartitions = listMatchingPartitionPaths(partitionFilters)
-    val listedPartitions = getInputFileSlices(prunedPartitions: _*).asScala.toSeq.map {
-      case (partition, fileSlices) =>
-        val baseFileStatuses: Seq[FileStatus] = getBaseFileStatus(fileSlices
-          .asScala
-          .map(fs => fs.getBaseFile.orElse(null))
-          .filter(_ != null))
+    var totalFileSliceSize = 0
+    var candidateFileSliceSize = 0
 
+    val listedPartitions = partitionAndFileSlices.map {
+      case (partitionOpt, fileSlices) =>
         // Filter in candidate files based on the col-stats index lookup
-        val candidateFiles = baseFileStatuses.filter(fs =>
-          // NOTE: This predicate is true when {@code Option} is empty
-          candidateFilesNamesOpt.forall(_.contains(fs.getPath.getName)))
+        val candidateFileSlices: Seq[FileSlice] = {
+          fileSlices.filter(fs => {
+            val baseFileStatusOpt = getBaseFileStatus(Option.apply(fs.getBaseFile.orElse(null)))
+            val logFiles = fs.getLogFiles.collect(Collectors.toSet[HoodieLogFile]).asScala.toSet[HoodieLogFile]
+            // NOTE: This predicate is true when {@code Option} is empty
+            if (candidateFilesNamesOpt.forall(files => baseFileStatusOpt.exists(f => files.contains(f.getPath.getName))
+              || files.intersect(logFiles.map(f => f.getPath.getName)).nonEmpty)) {
+              true
+            } else {
+              false
+            }
```

Comment:
let's make sure that this block is covered in a unit test (essentially a file slice with only log files).

Reply:
I tried with bucket index but it does not create only log files initially. I can only see parquet files getting created and then subsequent inserts create log files. Included the test still.

Reply:
Please try `INMEMORY` index type. I just did an upsert with quickstart data and in-memory index as below, and it generates only log files.
```
df.write.format("hudi").
  option("hoodie.insert.shuffle.parallelism", "1").
  option("hoodie.upsert.shuffle.parallelism", "1").
  option(TABLE_TYPE_OPT_KEY,"MERGE_ON_READ").
  option(PRECOMBINE_FIELD_OPT_KEY, "ts").
  option(RECORDKEY_FIELD_OPT_KEY, "uuid").
  option(PARTITIONPATH_FIELD_OPT_KEY, "partitionpath").
  option(TABLE_NAME, tableName).
  option(OPERATION_OPT_KEY, "upsert").
  option("hoodie.index.type", "INMEMORY").
  mode(Append).
  save(basePath)
```